File: iosApp/iosApp/Modules/NFTsCollection/NFTsCollectionPresenter.swift
// Created by web3d4v on 29/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum NFTsCollectionPresenterEvent {
    case dismiss
    case nftDetail(identifier: String)
}

protocol NFTsCollectionPresenter {
    func present()
    func handle(_ event: NFTsCollectionPresenterEvent)
}

final class DefaultNFTsCollectionPresenter {
    private weak var view: NFTsCollectionView!
    private let wireframe: NFTsCollectionWireframe
    private let interactor: NFTsCollectionInteractor
    private let context: NFTsCollectionWireframeContext

    private var latestNFTCollection: NFTCollection?
    private var latestNFTs: [NFTItem]?

    init(
        view: NFTsCollectionView,
        wireframe: NFTsCollectionWireframe,
        interactor: NFTsCollectionInteractor,
        context: NFTsCollectionWireframeContext
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.context = context
    }
}

extension DefaultNFTsCollectionPresenter: NFTsCollectionPresenter {

    func present() {
        interactor.fetchCollection(with: context.nftCollectionIdentifier) { [weak self] result in
            guard let self = self else { return }
            self.handleNFTCollectionResponse(with: result)
        }
        interactor.fetchNFTs(forCollection: context.nftCollectionIdentifier) { [weak self] result in
            guard let self = self else { return }
            self.handleNFTsResponse(with: result)
        }
    }

    func handle(_ event: NFTsCollectionPresenterEvent) {
        switch event {
        case .dismiss:
            wireframe.navigate(to: .dismiss)
        case let .nftDetail(identifier):
            wireframe.navigate(to: .nftDetail(identifier: identifier))
        }
    }
}

private extension DefaultNFTsCollectionPresenter {
    
    func handleNFTCollectionResponse(
        with result: Result<NFTCollection, Error>
    ) {
        switch result {
        case let .success(collection):
            latestNFTCollection = collection
            refreshView()
        case .failure:
            // We will handle failures in the future when connecting for real,
            // right now all data will be mocked and we know it won't return any errors.
            break
        }
    }
    
    func handleNFTsResponse(
        with result: Result<[NFTItem], Error>
    ) {
        switch result {
        case let .success(nfts):
            latestNFTs = nfts
            refreshView()
        case .failure:
            // We will handle failures in the future when connecting for real,
            // right now all data will be mocked and we know it won't return any errors.
            break
        }
    }
    
    func refreshView() {
        guard
            let latestNFTCollection = latestNFTCollection,
            let latestNFTs = latestNFTs
        else { return }
        view.update(
            with: .loaded(
                collection: latestNFTCollection,
                nfts: latestNFTs
            )
        )
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftsCollection/NFTsCollectionPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.nftsCollection

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3lib.utils.bgDispatcher
import com.sonsofcrypto.web3lib.utils.withUICxt
import com.sonsofcrypto.web3walletcore.modules.nftsCollection.NFTsCollectionPresenterEvent.Select
import com.sonsofcrypto.web3walletcore.modules.nftsCollection.NFTsCollectionWireframeDestination.Dismiss
import com.sonsofcrypto.web3walletcore.modules.nftsCollection.NFTsCollectionWireframeDestination.NFTDetail
import com.sonsofcrypto.web3walletcore.services.nfts.NFTCollection
import com.sonsofcrypto.web3walletcore.services.nfts.NFTItem
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

sealed class NFTsCollectionPresenterEvent {
    data class Select(val idx: Int): NFTsCollectionPresenterEvent()
    object Dismiss: NFTsCollectionPresenterEvent()
}

interface NFTsCollectionPresenter {
    fun present()
    fun handle(event: NFTsCollectionPresenterEvent)
}

class DefaultNFTsCollectionPresenter(
    private val view: WeakRef<NFTsCollectionView>,
    private val wireframe: NFTsCollectionWireframe,
    private val interactor: NFTsCollectionInteractor,
    private val context: NFTsCollectionWireframeContext,
): NFTsCollectionPresenter {
    private var collection: NFTCollection? = null
    private var nfts: List<NFTItem> = emptyList()

    override fun present() {
        view.get()?.update(NFTsCollectionViewModel.Loading)
        CoroutineScope(bgDispatcher).launch {
            interactor.clearCache()
            val fetchedCollection = interactor.collection(context.collectionId)
            val fetchedNfts = interactor.nfts(context.collectionId)
            withUICxt {
                collection = fetchedCollection
                nfts = fetchedNfts
                updateView()
            }
        }
    }

    override fun handle(event: NFTsCollectionPresenterEvent) = when (event) {
        is Select -> {
            wireframe.navigate(
                NFTDetail(context.collectionId, nfts[event.idx].identifier)
            )
        }
        is NFTsCollectionPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
    }

    private fun updateView() {
        view.get()?.update(
            NFTsCollectionViewModel.Loaded(
                NFTsCollectionViewModel.Collection(
                    collection?.identifier ?: context.collectionId,
                    collection?.coverImage,
                    collection?.title ?: context.collectionId,
                    collection?.author
                ),
                nfts.map {
                    NFTsCollectionViewModel.NFT(
                        it.identifier,
                        it.gatewayImageUrl,
                        it.gatewayPreviewImageUrl,
                        it.mimeType,
                        it.fallbackText
                    )
                }
            )
        )
    }
}